### ISSUE TITLE
How can I keep some parts of http messages from being mutated?

### ISSUE THREAD

**Issue by yoonsch217** [denandz/fuzzotron#13]
HTTP input example:

```
GET /api/v1/target HTTP/1.1
```

When I run fuzzotron with this testcase, the generator mutates from this text, and the generated inputs are like:

```
GET /api/v1/target HTTP/12321312
```
or
```
GGG /api/v1/target HTTP/1.1
```

However, I want to keep `GET` part and `HTTP/1.1` part from being mutated, and want only `/api/v1/target` to be mutated.

Is there a way to achieve this?

**Comment by denandz** [comment]
The best way to achieve this would be to modify `callback.c`. I just tweaked it to pass the testcase struct (402ce12a6ed3b4b9337f996a09a52b19cda19ad) so you can mess with that data directly. Here's a quick and dirty example that does kinda what you want (emphasis on the dirty, I'd suggest writing more robust code for your fuzzer)

In `callback.c`

```
#include <string.h>

#include "generator.h"
#include "util.h"

// Called after the socket is connected but before the test case is sent.
void callback_pre_send(int sock, testcase_t * testcase){
    // set up the template
    char * template = "GET  HTTP/1.1\r\n";
    char * packet;
    
    // alloc a new buffer and copy the template data in
    unsigned long len = strlen(template) + testcase->len;
    ft_malloc(len, packet);
    memset(packet, 0x00, len);
    
    memcpy(packet, template, strlen(template));
    memmove(packet+4+testcase->len, packet+4, 11);
    memcpy(packet+4, testcase->data, testcase->len);

    free(testcase->data); // free the original testcase
    testcase->data = packet; // update with new data
    testcase->len = len;
}
```

The testcases then just include the data you want fuzzed:

```
$ echo -n '/one/two/three' > cases/1
$ echo -n '/fooooo' > cases/2
$ ./fuzzotron --radamsa --directory cases -o foo -h 127.0.0.1 -p8181 -P tcp 
```

Which ends up sending...

```
GET /one\x0a!xcalc%n!xcalc%p%n\r\u0000'xcalc\x0a"xcalc\x127a\u0000%s%n\r\r\n`xcalc`$'/two/three HTTP/1.1
GET /one/two/three/one/two/three HTTP/1.1
GET /fooooo/fooooo HTTP/1.1
GET /fopoo󠁂o HTTP/1.1
GET שּׁ��fooopo HTTP/1.1
GET �⁥󠁥�/t�󠁑�hhree HTTP/1.1
GET /one/two/th�ree HTTP/1.1
GET 𝅘𝅥𝅮/fooooo HTTP/1.1
GET /oneee�w���/oneee HTTP/1.1
GET /foooooS/foooo󠀱o HTTP/1.1
GET ΐ�󠁷��������f��oooooooo HTTP/1.1
```

**Comment by denandz** [comment]
If this sounds like a huge pain in the ass, it's 'cause it is. If you just want to fuzz a specific part of an HTTP message and you don't care about the other features that Fuzzotron gives you, you could try using https://github.com/ffuf/ffuf with the `-request` flag for an easier option?

**Comment by denandz** [comment]
Resolved by f402ce1 - closing this issue
